### DISCUSSION TITLE
How to access elements in a 'hidden new window'?

### DISCUSSION THREAD

**Discussion by Angel-Dzhambazov** [cypress-io/cypress#29772] [Questions and Help]
Let's have for example a public page like https://www.sixt-neuwagen.de/ when we first open it we see that a cookie window comes up. This window does not look generic at all and it's hard for me to navigate to its elements. Inside of it we see a shadow-root. How do we access this new window?  Let's say for example, how do we accept the cookies for this website?

**Comment by portyu9** [comment]
Hi @Angel-Dzhambazov 

What you are describing sounds like **Shadow DOM**, not a separate browser window.

If the cookie-consent UI is inside an **open shadow root**, Cypress can traverse into it directly.

For example:

```js
cy.get('cookie-banner')
  .shadow()
  .find('button')
  .contains('Accept')
  .click()
```

For nested shadow roots, continue traversing:

```js
cy.get('outer-component')
  .shadow()
  .find('inner-component')
  .shadow()
  .find('button')
  .click()
```

You can also enable Shadow DOM traversal globally:

```js
import { defineConfig } from 'cypress'

export default defineConfig({
  includeShadowDom: true,
})
```

or use the `includeShadowDom` option on supported queries:

```js
cy.get('button', { includeShadowDom: true })
  .contains('Accept')
  .click()
```

I would first inspect the element in DevTools and identify the element immediately above the `#shadow-root (open)` entry. That is the **shadow host** to pass to `.shadow()`.

One limitation: if DevTools says:

```text
#shadow-root (closed)
```

then Cypress cannot traverse into it through the normal Shadow DOM API because the application intentionally does not expose that root.

Also check whether the consent component is additionally inside an `<iframe>`. A cross-origin iframe is a different problem from Shadow DOM and `.shadow()` alone would not cross that boundary.

Please mark this answer as accepted if it helped, thank you.
